### PR TITLE
feat(sdk): add erc20 client and table 

### PR DESCRIPTION
**Pull request by daywiss**
**Motivation**
https://app.clubhouse.io/uma-project/story/1296/add-erc20-client-and-table-to-sdk

**Summary**

Need to be able to get token decimals for TVL/TVM in API as well as token name.  Generalized this so that we can have a erc20 client with typed data and store the token information in a table.

**Details**

- adds sdk/clients/erc20
- adds sdk/tables/erc20
- some small refactors to emp/registry clients to make them more consistent

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [x]  Ran end-to-end test, running the code as in production
- [x]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

**Issue(s)**

https://app.clubhouse.io/uma-project/story/1296/add-erc20-client-and-table-to-sdk

### CODE REVIEW

File: packages/sdk/src/libs/clients/emp/client.ts [hunk 1]
```diff
@@ -16,13 +16,13 @@ export interface EventState {
...
-export type Redeem = GetEventType<ExpiringMultiParty, "Redeem">;
-export type Withdrawal = GetEventType<ExpiringMultiParty, "Withdrawal">;
-export type LiquidationCreated = GetEventType<ExpiringMultiParty, "LiquidationCreated">;
+export type RequestTransferPositionExecuted = GetEventType<Instance, "RequestTransferPositionExecuted">;
+export type PositionCreated = GetEventType<Instance, "PositionCreated">;
+export type NewSponsor = GetEventType<Instance, "NewSponsor">;
+export type SettleExpiredPosition = GetEventType<Instance, "SettleExpiredPosition">;
+export type Redeem = GetEventType<Instance, "Redeem">;
+export type Withdrawal = GetEventType<Instance, "Withdrawal">;
+export type LiquidationCreated = GetEventType<Instance, "LiquidationCreated">;
```
Comment (daywiss) [inline packages/sdk/src/libs/clients/emp/client.ts:25]: you can ignore this, just renaming vars to make it easier to make templates from these files

File: packages/sdk/src/libs/clients/erc20/client.test.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+require("dotenv").config();
+import assert from "assert";
+import * as Client from "./client";
+import { ethers } from "ethers";
+
+const address = "0xeca82185adCE47f39c684352B0439f030f860318";
+// these require integration testing, skip for ci
+describe.skip("erc20", function () {
```
Comment (chrismaree) [inline packages/sdk/src/libs/clients/erc20/client.test.ts:8]: would it be better to not use the sky syntax and have a different way to toggle if tests are run or not?

File: packages/sdk/src/libs/clients/erc20/README.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+const provider = new ethers.providers.WebSocketProvider(env.CUSTOM_NODE_URL)
+
+// get the contract instance
+const erc20Address:string = // assume you have an emp address you want to connect to
+const erc20Instance:uma.clients.erc20.Instance = uma.clients.erc20.connect(erc20Address,provider)
+// gets all emp events
+const events = await erc20Instance.queryFilter({})
+
+// returns EventState, defined in the emp client
+const state:uma.clients.erc20.EventState = uma.clients.erc20.getEventState(events)
```
Comment (chrismaree) [inline packages/sdk/src/libs/clients/erc20/README.md:21]: can this do more than just return events? should this also be documented?

**Review by chrismaree — APPROVED** [review]

File: packages/sdk/src/libs/clients/erc20/README.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+```js
+import {ethers} from 'ethers'
+import * as uma from '@uma/sdk'
+
+// assume you have a url injected from env
+const provider = new ethers.providers.WebSocketProvider(env.CUSTOM_NODE_URL)
+
+// get the contract instance
+const erc20Address:string = // assume you have an emp address you want to connect to
+const erc20Instance:uma.clients.erc20.Instance = uma.clients.erc20.connect(erc20Address,provider)
```
[the remark this answers is not in the document]
Comment (daywiss) [reply packages/sdk/src/libs/clients/erc20/README.md:16]: not sure i understand, what would be an example of this?

File: packages/sdk/src/libs/clients/erc20/README.md (hunk 3 above)
Comment (daywiss) [reply packages/sdk/src/libs/clients/erc20/README.md:21]: yea i should!

File: packages/sdk/src/libs/clients/erc20/client.test.ts (hunk 2 above)
Comment (daywiss) [reply packages/sdk/src/libs/clients/erc20/client.test.ts:8]: the right thing to do here is probably to add hard hat tests, but i've been kicking that can down the road. if you think that is probably the right course of action here i will make an issue for it

File: packages/sdk/src/libs/clients/erc20/README.md (hunk 3 above)
Comment (daywiss) [reply packages/sdk/src/libs/clients/erc20/README.md:22]: added more comments, but the client is pretty light, basicaly returns a contract instance, events and will decode events into some state

File: packages/sdk/src/libs/clients/erc20/client.test.ts (hunk 2 above)
Comment (daywiss) [reply packages/sdk/src/libs/clients/erc20/client.test.ts:8]: added a low priority card for this on clubhouse

**Pull request merged**